File: examples/fridgeclient.cpp
//******************************************************************
//
// Copyright 2014 Intel Mobile Communications GmbH All Rights Reserved.
//
//-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
//-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

/// This fridgeclient represents a client trying to discover the associated
/// fridgeserver.  The device resource is the only one available for discovery
/// on the server, so we have to take the fact that we know the device tag
/// to then generate a Resource object

#include <iostream>
#include <stdexcept>
#include <condition_variable>
#include <mutex>
#include "OCPlatform.h"
#include "OCApi.h"

using namespace OC;
namespace PH = std::placeholders;

// Option ID for API version and client token
const uint16_t API_VERSION = 2048;
const uint16_t TOKEN = 3000;

class ClientFridge
{
    public:
    ClientFridge(PlatformConfig &cfg) : m_platform(cfg)
    {
        std::cout << "Fridge Client has started " <<std::endl;
        FindCallback f (std::bind(&ClientFridge::foundDevice, this, PH::_1));

        OCStackResult result = m_platform.findResource(
                "", "coap://224.0.1.187/oc/core?rt=intel.fridge", f);

        if(OC_STACK_OK != result)
        {
            throw new std::runtime_error("Fridge Find Resource Failed");
        }

        std::cout << "Waiting to discover fridge... "<<std::endl;
        {
            // we want to block this thread until the client has finished
            // its duties, so we block on the CV until we have completed
            // what we are looking to do
            std::unique_lock<std::mutex> lk(m_mutex);
            m_cv.wait(lk);
        }
    }

    private:
    void foundDevice(std::shared_ptr<OCResource> resource)
    {
        if(resource && resource->uri() == "/device")
        {
            std::cout << "Discovered a device object"<<std::endl;
            std::cout << "\tHost: "<<resource->host()<<std::endl;
            std::cout << "\tURI:  "<<resource->uri() <<std::endl;
        }

        // we have now found a resource, so lets create a few resource objects
        // for the other resources that we KNOW are associated with the intel.fridge
        // server, and query them.
        std::vector<std::string> lightTypes = {"intel.fridge.light"};
        std::vector<std::string> ifaces = {DEFAULT_INTERFACE};
        OCResource::Ptr light = m_platform.constructResourceObject(resource->host(),
                                "/light", false, lightTypes, ifaces);

        std::vector<std::string> doorTypes = {"intel.fridge.door"};

        OCResource::Ptr leftdoor = m_platform.constructResourceObject(resource->host(),
                                "/door/left", false, doorTypes, ifaces);
        OCResource::Ptr rightdoor = m_platform.constructResourceObject(resource->host(),
                                "/door/right", false, doorTypes, ifaces);
        OCResource::Ptr randomdoor = m_platform.constructResourceObject(resource->host(),
                                "/door/random", false, doorTypes, ifaces);

        // Set header options with API version and token
        HeaderOptions headerOptions;
        try
        {
            // Set API version and client token
            HeaderOption::OCHeaderOption apiVersion(API_VERSION, "v.1.0");
            HeaderOption::OCHeaderOption clientToken(TOKEN, "21ae43gf");
            headerOptions.push_back(apiVersion);
            headerOptions.push_back(clientToken);
        }
        catch(OCException& e)
        {
            std::cout << "Error creating HeaderOption: " << e.what() << std::endl;
        }


        // Setting header options will send above options in all requests
        // Header options are set per resource.
        // Below, header options are set only for device resource
        resource->setHeaderOptions(headerOptions);

        resource->get(QueryParamsMap(), GetCallback(
                std::bind(&ClientFridge::getResponse, this, "Device", PH::_1,
                    PH::_2, PH::_3, resource, 0)
                ));
        light->get(QueryParamsMap(), GetCallback(
                std::bind(&ClientFridge::getResponse, this, "Fridge Light", PH::_1,
                    PH::_2, PH::_3, light, 1)
                ));
        leftdoor->get(QueryParamsMap(), GetCallback(
                std::bind(&ClientFridge::getResponse, this, "Left Door", PH::_1,
                    PH::_2, PH::_3, leftdoor, 2)
                ));
        rightdoor->get(QueryParamsMap(), GetCallback(
                std::bind(&ClientFridge::getResponse, this, "Right Door", PH::_1,
                    PH::_2, PH::_3, rightdoor, 3)
                ));
        randomdoor->get(QueryParamsMap(), GetCallback(
                std::bind(&ClientFridge::getResponse, this, "Random Door", PH::_1,
                    PH::_2, PH::_3, randomdoor, 4)
                ));
    }

    // Note that resourceName, resource, and getId are all bound via the std::bind mechanism.
    // it is possible to attach ANY arbitrary data to do whatever you would like here.  It may,
    // however be a better fit to wrap each call in an object so a fuller context (and additional
    // requests) can be easily made inside of a simple context
    void getResponse(const std::string& resourceName, const HeaderOptions& headerOptions,
                const OCRepresentation rep, const int eCode, OCResource::Ptr resource, int getId)
    {
        std::cout << "Got a response from get from the "<<resourceName<< std::endl;
        std::cout << "Get ID is "<<getId<<" and resource URI is "<<resource->uri()<<std::endl;

        std::cout << "The Attribute Data is: "<<std::endl;

        for (auto it = headerOptions.begin(); it != headerOptions.end(); ++it)
        {
            if(it->getOptionID() == API_VERSION)
            {
                std::cout << "Server API version in GET response: " <<
                        it->getOptionData() << std::endl;
            }
        }

        switch(getId)
        {
            case 0:
                {
                    // Get on device
                    std::string name;
                    rep.getValue("device_name", name);
                    std::cout << "Name of device: "<< name << std::endl;
                    break;
                }
            case 1:
                {
                    bool isOn;
                    rep.getValue("on",isOn);
                    std::cout<<"The fridge light is "<< ((isOn)?"":"not ") <<"on"<<std::endl;
                }
                break;
            case 2:
            case 3:
                {
                    bool isOpen;
                    std::string side;
                    rep.getValue("open", isOpen);
                    rep.getValue("side", side);
                    std::cout << "Door is "<<isOpen<<" and is on the "<<side<<std::endl;
                }
                break;
            case 4:
                {
                    // Get on random resource called.
                    std::string name;
                    rep.getValue("device_name", name);
                    std::cout << "Name of fridge: "<< name << std::endl;
                    break;
                }
        }
    }

    OCPlatform m_platform;
    std::mutex m_mutex;
    std::condition_variable m_cv;
};

int main()
{
    PlatformConfig cfg
    {
        ServiceType::InProc,
        ModeType::Client,
        "0.0.0.0",
        0,
        QualityOfService::LowQos
    };

    ClientFridge cf (cfg);
    return 0;
}

File: android/examples/fridgeclient/src/main/java/org/iotivity/base/examples/fridgeclient/FridgeClient.java
/*
 * //******************************************************************
 * //
 * // Copyright 2015 Intel Corporation.
 * //
 * //-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
 * //
 * // Licensed under the Apache License, Version 2.0 (the "License");
 * // you may not use this file except in compliance with the License.
 * // You may obtain a copy of the License at
 * //
 * //      http://www.apache.org/licenses/LICENSE-2.0
 * //
 * // Unless required by applicable law or agreed to in writing, software
 * // distributed under the License is distributed on an "AS IS" BASIS,
 * // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * // See the License for the specific language governing permissions and
 * // limitations under the License.
 * //
 * //-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
 */

package org.iotivity.base.examples.fridgeclient;

import android.app.Activity;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.os.Message;
import android.support.v4.content.LocalBroadcastManager;
import android.text.method.ScrollingMovementMethod;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.LinearLayout;
import android.widget.TextView;

import org.iotivity.base.ErrorCode;
import org.iotivity.base.ModeType;
import org.iotivity.base.OcConnectivityType;
import org.iotivity.base.OcException;
import org.iotivity.base.OcHeaderOption;
import org.iotivity.base.OcPlatform;
import org.iotivity.base.OcRepresentation;
import org.iotivity.base.OcResource;
import org.iotivity.base.PlatformConfig;
import org.iotivity.base.QualityOfService;
import org.iotivity.base.ServiceType;

import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;

import base.iotivity.org.examples.message.IMessageLogger;

/**
 * FridgeClient
 * <p/>
 * FridgeClient is a sample client app which should be started after the fridgeServer is started.
 * It creates DeviceResource, DoorResource, LightResource and performs a get operation on them.
 * This implements IMessageLogger to display messages on the screen
 */
public class FridgeClient extends Activity implements
        OcPlatform.OnResourceFoundListener, IMessageLogger {
    private static String TAG = "FridgeClient: ";

    private MessageReceiver mMessageReceiver = new MessageReceiver();
    private TextView mEventsTextView;
    private String mDeviceName;
    private int mDeviceCode;
    private List<String> ifaces;
    private final List<OcResource> resourceList = new LinkedList<OcResource>();

    /**
     * configure OIC platform and call findResource
     */
    private void initOICStack() {
        PlatformConfig cfg = new PlatformConfig(
                this,
                ServiceType.IN_PROC,
                ModeType.CLIENT,
                "0.0.0.0", // bind to all available interfaces
                0,
                QualityOfService.LOW);

        OcPlatform.Configure(cfg);
        try {
            OcPlatform.findResource("", OcPlatform.WELL_KNOWN_QUERY + "?rt=" + "intel.fridge",
                    OcConnectivityType.WIFI, this);
        } catch (OcException e) {
            logMessage(TAG + " init Error. " + e.getMessage());
            Log.e(TAG, e.getMessage());
        }
    }

    /**
     * prints out the appropriate messages depending on the device code
     *
     * @param representation representation of the OcResource
     * @param value          clientDeviceCode
     */
    private void getResponse(OcRepresentation representation, int value) {
        switch (value) {
            case 0:
                // Get on device
                try {
                    logMessage(TAG + "Name of device: " +
                            representation.getValue(StringConstants.DEVICE_NAME));
                } catch (OcException e) {
                    Log.e(TAG, e.getMessage());
                }
                break;
            case 1:
                // get on fridge light
                try {
                    boolean lightOn = representation.getValue(StringConstants.ON);
                    logMessage(TAG + "The fridge light is " +
                            (lightOn ? "" : "not " + "on"));
                } catch (OcException e) {
                    Log.e(TAG, e.getMessage());
                }
                break;
            case 2:
            case 3:
                // get on fridge door(s)
                try {
                    boolean doorOpen = representation.getValue(StringConstants.OPEN);
                    logMessage(TAG + "Door is " + (doorOpen ?
                            "open" : "not open") + " and is on the " +
                            representation.getValue(StringConstants.SIDE) + " side");
                } catch (OcException e) {
                    Log.e(TAG, e.getMessage());
                }
                break;
            case 4:
                // get on fridge random door
                try {
                    logMessage("Name of fridge: " +
                            representation.getValue(StringConstants.DEVICE_NAME));
                } catch (OcException e) {
                    Log.e(TAG, e.getMessage());
                }
                break;
            default:
                logMessage("Unexpected State");
                break;
        }
    }

    /**
     * this method is used to wait for 1 second between calls to different resources.
     * It is added for better readability
     */
    private void doWait() {
        try {
            Thread.sleep(StringConstants.WAIT_TIME);
        } catch (InterruptedException e) {
            logMessage(TAG + "doWait exception: " + e.getMessage());
            Log.e(TAG, e.getMessage());
        }
    }

    @Override
    /**
     *  callback when a resource is found. This method calls getResponse with the correct code
     */
    synchronized public void onResourceFound(OcResource ocResource) {
        // eventHandler for onGetListener
        resourceList.add(ocResource);
        OcResource.OnGetListener onGetListener = new OcResource.OnGetListener() {
            @Override
            public void onGetCompleted(List<OcHeaderOption> headerOptionList, OcRepresentation rep) {
                logMessage(TAG + " Got a response from " + getClientDeviceName());
                getResponse(rep, getClientDeviceCode());
            }

            @Override
            public void onGetFailed(Throwable throwable) {
                if (throwable instanceof OcException) {
                    OcException ocEx = (OcException) throwable;
                    ErrorCode errCode = ocEx.getErrorCode();
                    //do something based on errorCode
                }
                Log.e(TAG, throwable.toString());
            }
        };

        if (ocResource.getUri().equals(StringConstants.RESOURCE_URI)) {
            logMessage(TAG + "Discovered a device with \nHost: " + ocResource.getHost() +
                    ", Uri: " + ocResource.getUri());
        }
        List<String> lightTypes = new LinkedList<>();
        lightTypes.add("intel.fridge.light");
        try {
            OcResource light = OcPlatform.constructResourceObject(ocResource.getHost(),
                    StringConstants.LIGHT, OcConnectivityType.WIFI, false, lightTypes, ifaces);

            List<String> doorTypes = new LinkedList<>();
            doorTypes.add("intel.fridge.door");
            OcResource leftDoor = OcPlatform.constructResourceObject(ocResource.getHost(),
                    StringConstants.LEFT_DOOR, OcConnectivityType.WIFI, false, doorTypes, ifaces);

            OcResource rightDoor = OcPlatform.constructResourceObject(ocResource.getHost(),
                    StringConstants.RIGHT_DOOR, OcConnectivityType.WIFI, false, doorTypes, ifaces);

            OcResource randomDoor = OcPlatform.constructResourceObject(ocResource.getHost(),
                    StringConstants.RANDOM_DOOR, OcConnectivityType.WIFI, false, doorTypes, ifaces);

            List<OcHeaderOption> headerOptions = new LinkedList<>();
            OcHeaderOption apiVersion = new OcHeaderOption(StringConstants.API_VERSION_KEY,
                    StringConstants.API_VERSION);
            OcHeaderOption clientToken = new OcHeaderOption(StringConstants.CLIENT_VERSION_KEY,
                    StringConstants.CLIENT_TOKEN);
            headerOptions.add(apiVersion);
            headerOptions.add(clientToken);
            ocResource.setHeaderOptions(headerOptions);
            /**
             *  wait for 1 second before calling get on different resources.
             *  It is done for better readability.
             *  doWait() is called before each call to get
             */
            doWait();

            setupClientOptions("Device", 0);
            ocResource.get(new HashMap<String, String>(), onGetListener);
            doWait();

            setupClientOptions("Fridge Light", 1);
            light.get(new HashMap<String, String>(), onGetListener);
            doWait();

            setupClientOptions("Left Door", 2);
            leftDoor.get(new HashMap<String, String>(), onGetListener);
            doWait();

            setupClientOptions("Right Door", 3);
            rightDoor.get(new HashMap<String, String>(), onGetListener);
            doWait();

            setupClientOptions("Random Door", 4);
            randomDoor.get(new HashMap<String, String>(), onGetListener);
            doWait();

            resourceList.add(leftDoor);
            leftDoor.deleteResource(new OcResource.OnDeleteListener() {
                @Override
                public void onDeleteCompleted(List<OcHeaderOption> ocHeaderOptions) {
                    logMessage(TAG + "Delete resource successful");
                }

                @Override
                public void onDeleteFailed(Throwable throwable) {
                    if (throwable instanceof OcException) {
                        OcException ocEx = (OcException) throwable;
                        ErrorCode errCode = ocEx.getErrorCode();
                        //do something based on errorCode
                    }
                    Log.e(TAG, throwable.toString());
                }
            });
        } catch (OcException e) {
            logMessage(TAG + "onResourceFound Error. " + e.getMessage());
            Log.e(TAG, e.getMessage());
        }
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_fridge_client);
        registerReceiver(mMessageReceiver, new IntentFilter(StringConstants.INTENT));

        mEventsTextView = new TextView(this);
        mEventsTextView.setMovementMethod(new ScrollingMovementMethod());
        LinearLayout layout = (LinearLayout) findViewById(R.id.linearLayout);
        layout.addView(mEventsTextView, new LinearLayout.LayoutParams
                (LinearLayout.LayoutParams.MATCH_PARENT, 0, 1f));
        ifaces = new LinkedList<>();
        ifaces.add(StringConstants.RESOURCE_INTERFACE);
        mDeviceCode = -1;
        mDeviceName = "";

        initOICStack();
    }

    public class MessageReceiver extends BroadcastReceiver {
        @Override
        public void onReceive(Context context, Intent intent) {
            final String message = intent.getStringExtra(StringConstants.MESSAGE);
            logMessage(message);
        }
    }

    @Override
    public void logMessage(final String text) {
        if (StringConstants.ENABLE_PRINTING) {
            runOnUiThread(new Runnable() {
                public void run() {
                    final Message msg = new Message();
                    msg.obj = text;
                    mEventsTextView.append("\n");
                    mEventsTextView.append(text);
                }
            });
            Log.i(TAG, text);
        }
    }


    private void setupClientOptions(String name, int value) {
        mDeviceName = name;
        mDeviceCode = value;
    }

    private String getClientDeviceName() {
        return mDeviceName;
    }

    private int getClientDeviceCode() {
        return mDeviceCode;
    }


    //method to print the headerOptions received from the server
    void printHeaderOptions(List<OcHeaderOption> headerOptions) {
        for (OcHeaderOption headerOption : headerOptions) {
            if (StringConstants.API_VERSION_KEY == headerOption.getOptionId()) {
                logMessage(TAG + "Server API version in GET response: " +
                        headerOption.getOptionData());
            }
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_fridge_client, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        if (id == R.id.action_settings) {
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        onStop();
    }

    @Override
    protected void onStop() {
        LocalBroadcastManager.getInstance(this).unregisterReceiver(mMessageReceiver);
        super.onStop();
    }
}
